pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/store/FilePerIndexDirectory.java

```diff
@@ -88,26 +91,42 @@ public boolean hasIndexFor(String column, ColumnIndexType type) {
   @Override
   public void close()
       throws IOException {
-    for (PinotDataBuffer dataBuffer : indexBuffers.values()) {
+    for (PinotDataBuffer dataBuffer : _indexBuffers.values()) {
       dataBuffer.close();
     }
   }
 
   @Override
   public void removeIndex(String columnName, ColumnIndexType indexType) {
     File indexFile = getFileFor(columnName, indexType);
-    indexFile.delete();
+    if (indexFile.delete()) {
+      _indexBuffers.remove(new IndexKey(columnName, indexType));
+    }
   }
 
   @Override
   public boolean isIndexRemovalSupported() {
     return true;
   }
 
+  @Override
+  public Set<String> getColumnsWithIndex(ColumnIndexType type) {
+    if (_indexBuffers.isEmpty()) {
+      return Collections.emptySet();
+    }
```

Comment:
I don't think this check is required as it should be super rare that there is no index in the segment